opentelemetry/api/trace/__init__.py

```diff
@@ -0,0 +1,241 @@
+    with tracer.start_span("parent"):
+        # Attach a new child and update the current span
+        with tracer.start_span("child"):
+            do_work():
+        # Close child span, set parent as current
+    # Close parent span, set default span as current
+
+When creating a span that's "detached" from the context the active span doesn't
+change, and the caller is responsible for managing the span's lifetime::
+
+    from opentelemetry.sdk.trace import tracer
+
+    # Explicit parent span assignment
+    span = tracer.create_span("child", parent=parent) as child:
+
+    # The caller is responsible for starting and ending the span
+    span.start()
+    try:
+        do_work(span=child)
+    finally:
+        span.end()
+
+Applications should generally use a single global tracer, and use either
+implicit or explicit context propagation consistently throughout.
+
+.. versionadded:: 0.1.0
+"""
+
+from __future__ import annotations
+
+from contextlib import contextmanager
+from typing import Iterator
+
+
+class Tracer(object):
+    """Handles span creation and in-process context propagation.
+
+    This class provides methods for manipulating the context, creating spans,
+    and controlling spans' lifecycles.
+    """
+
+    def get_current_span(self) -> Span:
+        """Gets the currently active span from the context.
+
+        If there is no current span, return a placeholder span with an invalid
+        context.
+
+        Returns:
+            The currently active :class:`.Span`, or a placeholder span with an
+            invalid :class:`.SpanContext`.
+        """
+        raise NotImplementedError
+
+
+    @contextmanager
+    def start_span(self, name: str, parent: Span) -> Iterator[Span]:
+        """Context manager for span creation.
+
+        Create a new child of the current span, or create a root span if no
+        current span exists. Start the span and set it as the current span in
+        this tracer's context.
+
+        On exiting the context manager stop the span and set its parent as the
+        current span.
+
+        Example::
+
+            with tracer.span("one") as parent:
+                parent.add_event("parent's event")
+                with tracer.span("two") as child:
+                    child.add_event("child's event")
+                    tracer.get_current_span()  # returns child
+                tracer.get_current_span()      # returns parent
+            tracer.get_current_span()          # returns the previously active span
+
+        This is a convenience method for creating spans attached to the
+        tracer's context. Applications that need more control over the span
+        lifetime should use :meth:`create_span` instead. For example::
+
+            with tracer.start_span(name) as span:
+                do_work()
+
+        is equivalent to::
+
+            span = tracer.create_span(name, parent=tracer.get_current_span())
+            with tracer.use_span(span):
```

Comment:
Maybe I've missed something, I was thinking a different approach:

##### Implicit
```python
with tracer.span(name):
    do_work()
```

##### Explicit
```python
span = tracer.span(name)
span.begin()
do_work()
span.end()
```

And in the span class we will have `__enter__` and `__exit__` (or `__aenter__` and `__aexit__`).

Reply:
That's a nice interface, but what do you do about the fact that we need to modify the context if span is used as a context manager and not otherwise?